### PR TITLE
Feature: Identify Confounders Utility Function

### PR DESCRIPTION
**Pull request by steven-causalens**
## Description and Motivation
This PR adds the `identify_confounders` utility function, which allows you to identify a list of confounders between two nodes within a causal graph.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [x] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [x] I have implemented all requirements? (see JIRA, project documentation)
- [x] I am affecting someone else's work? If so: I have added those people as reviewers?
- [x] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [x] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [x] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [x] If needed, documentation has been added/updated?
- [x] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

**Comment by max-causalens** [comment]
It would be helpful to make the user of this package aware that these utilities exist, they're very valuable but without being linked anywhere in the narrative docs, people will not know they exist.

**Review by max-causalens — APPROVED** [review]

File: cai_causal_graph/identify_utils.py
Comment (andrew-causalens) [inline cai_causal_graph/identify_utils.py:1]: Would actually just put into `utils`. To Max's point about floating these exist, would suggesting add a new markdown that just describes the identifying utils. Can ignore the exists utils as they aren't too exiting.

File: cai_causal_graph/identify_utils.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+"""
+Copyright (c) 2023 by Impulse Innovations Ltd. Private and confidential. Part of the causaLens product.
+"""
+from typing import List
+
+from cai_causal_graph import CausalGraph
+from cai_causal_graph.exceptions import CausalGraphErrors
+
+
+def identify_confounders(graph: CausalGraph, treatment: str, outcome: str) -> List[str]:
```
Comment (andrew-causalens) [inline cai_causal_graph/identify_utils.py:10]: Confused by the note as it was missing a node: 

"
Note: The set of confounders is must be a sufficient adjustment set but not every adjustment set also constitutes a set of confounders. As example, consider this graph: `z -> u -> x ->` and `z -> y.` 

The confounder between `x` and `y` is clearly `z`, but `u` is actually also a sufficient (minimal) adjustment set. Since all causal effect estimators have the wording confounders = ... in their constructors, I believe that this is fine. Let me know if you believe otherwise.
"

The causal graph package should not care about effect estimation per say. I think the util function should give the correct answer for the graph. But I was not sure what the note was saying. Are we actually providing more nodes than the truth?

File: cai_causal_graph/identify_utils.py
Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py:1]: I agree with @max-causalens and you, and I have already added a section to the `causal_graph.md` docs file. Are you suggesting to create a new markdown file?

Comment (andrew-causalens) [reply cai_causal_graph/identify_utils.py:1]: Yeah because these are util functions and don't live on `CausalGraph`.

File: cai_causal_graph/identify_utils.py (hunk 1 above)
Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py]: Ah apologies, I was missing a `y` in the nodes.

All I was saying is that:
- A set of confounders is a sufficient adjustment set.
- An adjustment set, however, does not need to constitute a set of confounders.

I agree that we do not really care in the `cai-causal-graph` package, and I have modified the argument names to reflect that too.

File: cai_causal_graph/identify_utils.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+"""
+Copyright (c) 2023 by Impulse Innovations Ltd. Private and confidential. Part of the causaLens product.
```
Comment (andrew-causalens) [inline cai_causal_graph/identify_utils.py:2]: Not the correct header for open source code.

File: cai_causal_graph/identify_utils.py (hunk 1 above)
Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py]: Updated the PR description as well.

File: tests/test_identify_utils.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+"""
+Copyright (c) 2023 by Impulse Innovations Ltd. Private and confidential. Part of the causaLens product.
```
Comment (andrew-causalens) [inline tests/test_identify_utils.py:2]: Same here. Wrong header.

File: tests/test_identify_utils.py
Comment (andrew-causalens) [inline tests/test_identify_utils.py:1]: Add one test for non DAG to show it raises appropriate error.

File: cai_causal_graph/identify_utils.py
Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py:1]: FYI, I cannot add this function into `utils.py` because of circular import errors. I have to have another file.

File: cai_causal_graph/identify_utils.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    :param destination: The destination variable.
+    :return: A list of all confounders between the `source` and `destination`.
+    """
+    if not graph.is_dag():
+        raise TypeError(f'Expected a DAG, but got a mixed causal graph.')
+
+    # create a copy of the provided graph and prune the edge between source and destination
+    pruned_graph = graph.copy()
+    try:
+        pruned_graph.remove_edge(source=source, destination=destination)
```
Comment (andrew-causalens) [inline cai_causal_graph/identify_utils.py:58]: This ignores the edge type. What if we have Z -> X, Z -> Y, Y-X but say source=X, destination=Y then should we have empty list returned? However, this would return Z.

Comment (andrew-causalens) [reply cai_causal_graph/identify_utils.py:58]: Maybe we don't actually care source vs destination but just between 2 nodes.

File: changelog.md [hunk 5]
```diff
@@ -1,5 +1,10 @@
 # Changelog
 
+## NEXT
+
+- Added the `cai_causal_graph.identify_utils.identify_confounders` utility function, which allows you to identify
+  a list of confounders between two nodes in a causal graph.
```
Comment (andrew-causalens) [inline changelog.md:6]: "in a `cai_causal_graph.causal_graph.CausalGraph`".

File: docs/causal_graph.md [hunk 6]
```diff
@@ -341,3 +341,35 @@ causal_graph_dict: dict = causal_graph.to_dict()
 # Dictionary to CausalGraph
 causal_graph: CausalGraph = CausalGraph.from_dict(causal_graph_dict)
 ```
+
+## Identify Useful Subsets of a Causal Graph
```
Comment (andrew-causalens) [inline docs/causal_graph.md:345]: Do you want to keep this?

File: tests/test_identify_utils.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import unittest
+
+from cai_causal_graph import CausalGraph
+from cai_causal_graph.identify_utils import identify_confounders
+from cai_causal_graph.type_definitions import EDGE_T
+
+
+class TestIdentifyConfounders(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
```
Comment (andrew-causalens) [inline tests/test_identify_utils.py:25]: Suggest adding one test that it works with `TimeSeriesCausalGraph` as well. Should be fine as it is a subclass but worth exercising.

File: cai_causal_graph/identify_utils.py (hunk 4 above)
Comment (andrew-causalens) [reply cai_causal_graph/identify_utils.py:58]: If order of 2 nodes doesn't matter then would add unit test to confirm `identify_confounders(g, 'x', 'y')` and `identify_confounders(g, 'y', 'x')` returns the same list.

File: docs/causal_graph.md (hunk 6 above)
Comment (steven-causalens) [reply docs/causal_graph.md:345]: Nope, thanks for catching!

File: cai_causal_graph/identify_utils.py (hunk 4 above)
Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py:58]: I think that's a good suggestion. I do agree that we should get the same result if the `source` and `destination` are swapped. Cheers

Comment (andrew-causalens) [reply cai_causal_graph/identify_utils.py:58]: Should we change the name though as there is no real idea of source vs destination? It implies one is above other in topological sort. Would maybe just say `node_1` and `node_2`.

Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py:58]: I think just `source` and `destination` is totally fine to be honest, but I'm happy to change it if you have a strong opinion

Comment (mmiklitz) [reply cai_causal_graph/identify_utils.py:58]: I would be worried about the backward compatibility here. Maybe enough to explain in the docstring that the notion of source and destination is vague and maybe sometimes ignored in directed edges?

Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py:58]: If you believe it's necessary to explain this further in the docstrings, I'd rather just change the names to `node_1` and `node_2`

Comment (steven-causalens) [reply cai_causal_graph/identify_utils.py]: Changed now!

Comment (andrew-causalens) [reply cai_causal_graph/identify_utils.py]: Thanks! @mmiklitz no concern for backwards compat as this util never existed before.

File: tests/test_identify_utils.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        graph = CausalGraph()
+        graph.add_edge('z', 'u', edge_type=EDGE_T.UNDIRECTED_EDGE)
+        graph.add_edge('z', 'y')
+        graph.add_edge('u', 'x')
+        graph.add_edge('x', 'y')
+
+        with self.assertRaises(TypeError):
+            identify_confounders(graph, node_1='x', node_2='y')
+
+    def test_symmetric(self):
```
Comment (andrew-causalens) [inline tests/test_identify_utils.py:92]: @steven-causalens this test now fails so it seems like the function is not symmetric.

Comment (steven-causalens) [reply tests/test_identify_utils.py:92]: Thanks again for catching this @andrew-causalens, I've now extended the method. Essentially, I'm calling the reverse of the search to ensure symmetry and then taking the intersection to obtain the minimal set of confounders.

**Review by mmiklitz — APPROVED** [review]
Neat @steven-causalens!

File: cai_causal_graph/identify_utils.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    for parent in pruned_graph.get_parents(node_2):
+        # add the parent to the confounding set if a directed path exists
+        if parent in pruned_graph.get_ancestors(node_1):
+            confounders_reverse.add(parent)
+        # otherwise, recursively call this function to identify confounders of the parent
+        else:
+            confounders_reverse = confounders_reverse.union(set(identify_confounders(pruned_graph, parent, node_1)))
+
+    # take the intersection of both sets to get the minimal confounders
+    minimal_confounders = confounders.intersection(confounders_reverse)
```
Comment (andrew-causalens) [inline cai_causal_graph/identify_utils.py:81]: Can you just extend the comment why the intersection is the "correct" choice over union?

File: cai_causal_graph/identify_utils.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+from cai_causal_graph import CausalGraph
+
+
+def identify_confounders(graph: CausalGraph, node_1: str, node_2: str) -> List[str]:
+    """
+    Identify all confounders between `node_1` and `node_2` in the provided `graph`.
+
+    A confounder between `node_1` and `node_2` is a node that is a (minimal) ancestor of both the `node_1` and
+    `node_2`. Being a _minimal_ ancestor here means that the node is not an ancestor of other confounder nodes, unless
+    it has another directed path to `node_2` that does not go through other confounder nodes.
```
Comment (andrew-causalens) [inline cai_causal_graph/identify_utils.py:27]: Does this need slight refinement after updating it as it does both ways?

**Review by andrew-causalens — APPROVED** [review]
